### ISSUE TITLE
Store file sync state in project root metadata attrib to avoid data loss, deal with 503 timeouts, etc

### ISSUE THREAD

**Issue by humphd** [mozilla/thimble.webmaker.org#1115]
We see a lot of timeouts on requests we send.  Partly this is because we send a lot, and partly it's because of [issues in the login pipeline](https://github.com/mozilla/thimble.webmaker.org/issues/1112); there might be other reasons, too.

The net result is that we can and do see a lot of 503s doing things like this:
```
GET /projects/*
Sep 14 18:14:05 bramble-production heroku/router: at=error code=H13 desc="Connection closed without response" method=GET path="/projects/new?cacheBust=1442279574926" host=bramble-production.herokuapp.com request_id=ef8cb052-861e-470f-b58a-b2d69974818d fwd="72.182.83.115,216.137.42.89" dyno=web.2 connect=0ms service=9624ms status=503 bytes=0

PUT /projects/*
Sep 14 18:14:05 bramble-production heroku/router: at=error code=H13 desc="Connection closed without response" method=PUT path="/projects/9410/files/32484" host=bramble-production.herokuapp.com request_id=989b632b-d41b-42e4-b94c-1e3ff128fd25 fwd="174.49.44.113,216.137.42.113" dyno=web.2 connect=1ms service=19427ms status=503 bytes=0 

DELETE /projects/*
Sep 14 18:48:21 bramble-production heroku/router: at=error code=H12 desc="Request timeout" method=DELETE path="/projects/9261" host=bramble-production.herokuapp.com request_id=cf970d9c-aa37-41c0-8d14-adc40028fea9 fwd="172.56.11.94,205.251.252.98" dyno=web.2 connect=1ms service=30010ms status=503 bytes=0
```

We need to not have these timeouts happen so often, yes, but we should also build in some resilience so that the browser (and Thimble server) can survive timeouts from publish, login, and id.  The client should maybe done one or more of the following:
* try again
* wait, and try again
* cache the request somewhere stable so it can be re-run later

Our queue of file and project updates should probably not live in memory but in the filesystem, which is what #1092 is about.  It would be good if change events of one kind or another started by getting queued in the filesystem, then another service could read those and try to execute them, removing them from the queue when successful.  In this way we could restart uploads even after a reload.  If we were really smart, we could restart uploads for any project across the entire filesystem, no matter what the user opens (i.e., the queue could be global vs. per project, and keyed on absolute paths).

There are cases where we might want to tell the user that something is up, too.  For example, if you're trying to publish, and waiting for a URL, we should let you know that things are taking a long time, and we'll give you the URL in some notification later.

**Comment by humphd** [comment]
@gideonthomas and I discussed this today, and I think what I'm going to try is the following:

* we store paths and operations for those paths on the project root as an xattrib.  The data will look something like this:
```js
{
  "/path/to/file1" : ['update', 'update', 'update'],
  "/path/to/file2": ['update', 'update', 'delete'],
  ...
}
```
Here we have an object keyed on file paths.  Each file path has an array value, which records linear file operations to be applied, based on events from Bramble (autosaves, renames, deletes).
* we then have a service that runs on a schedule, or when the user clicks "Publish" and it drains this queue one by one.  It does this by randomly picking a path, and applying the smallest set of operations necessary to sync it with the server.  If there are multiple updates, they can be compressed into one update.  If there is a delete, the updates can be before it can be ignored, etc.
* if the serve responds with a 503 (timeout) or some other error, we leave the file in the queue, and try another one.  This way we don't lose track of state that we tried to sync and failed.
